extensions/big-data-cluster/src/bigDataCluster/tree/controllerTreeNode.ts

```diff
@@ -0,0 +1,376 @@
+				this.addSqlMasterNode(master.endpoint, master.description);
+			}
+			return this.children as ControllerTreeNode[];
+		} catch (error) {
+			let e = error as IControllerError;
+			vscode.window.showErrorMessage(e.message);
+			return this.children as ControllerTreeNode[];
+		}
+	}
+
+	private static toIpAndPort(url: string): string {
+		if (!url) {
+			return;
+		}
+		return url.trim().replace(/ /g, '').replace(/^.+\:\/\//, '').replace(/:(\d+)$/, ',$1');
+	}
+
+	public addSqlMasterNode(endPointAddress: string, description: string): void {
+		let epFolder = this.getEndPointFolderNode();
+		epFolder.addChild(new SqlMasterNode({ endPointAddress, parent: epFolder, treeChangeHandler: this.treeChangeHandler, description }));
+	}
+
+	private getEndPointFolderNode(): FolderNode {
+		let label = localize('textSqlServers', 'SQL Servers');
+		let epFolderNode = this.children.find(e => e instanceof FolderNode && e.label === label);
+		if (!epFolderNode) {
+			epFolderNode = new FolderNode({ label, parent: this, treeChangeHandler: this.treeChangeHandler });
+			this.addChild(epFolderNode);
+		}
+		return epFolderNode as FolderNode;
+	}
+
+	public getTreeItem(): vscode.TreeItem {
+		let item: vscode.TreeItem = super.getTreeItem();
+		item.collapsibleState = vscode.TreeItemCollapsibleState.Collapsed;
+		return item;
+	}
+
+	public get url() {
+		return this._url;
+	}
+
+	public set url(url: string) {
+		this._url = url;
+	}
+
+	public get username() {
+		return this._username;
+	}
+
+	public set username(username: string) {
+		this._username = username;
+	}
+
+	public get password() {
+		return this._password;
+	}
+
+	public set password(pw: string) {
+		this._password = pw;
+	}
+
+	public get rememberPassword() {
+		return this._rememberPassword;
+	}
+
+	public set rememberPassword(rememberPassword: boolean) {
+		this._rememberPassword = rememberPassword;
+	}
+}
+
+export type FolderNodeArg = {
+	label: string;
+	parent: ControllerTreeNode;
+	treeChangeHandler: IControllerTreeChangeHandler;
+};
+
+export class FolderNode extends ControllerTreeNode {
+	constructor(arg?: FolderNodeArg) {
+		super(Object.assign({
+			description: arg.label,
+			nodeType: BdcItemType.folder,
+			iconPath: IconPath.folderNode
+		}, arg));
+	}
+}
+
+export type SqlMasterNodeArg = {
+	endPointAddress: string,
+	parent: ControllerTreeNode,
+	treeChangeHandler: IControllerTreeChangeHandler,
+	label?: string,
+	description?: string,
+};
+
+export class SqlMasterNode extends ControllerTreeNode {
+	private _role: string;
+	private _endPointAddress: string;
+	private _username: string;
+	private _password: string;
+
+	constructor(arg?: SqlMasterNodeArg) {
+		super(Object.assign({
+			label: undefined,
+			nodeType: BdcItemType.sqlMaster,
+			iconPath: IconPath.sqlMasterNode
+		}, arg));
+
+		this._endPointAddress = arg.endPointAddress;
+		this._username = 'sa';
+		this._role = 'sql-server-master';
+		this.label = arg.label || `master: ${this._endPointAddress} (${this._username})`;
```

Comment:
this `master:` syntax should be localized? Is there a reason we need it at all? Perhaps run this by UX (no need to block on this).

Reply:
being tracked for UX review